mcstatus/forge_data.py

```diff
@@ -0,0 +1,183 @@
+
+For more information see this file from forge itself:
+https://github.com/MinecraftForge/MinecraftForge/blob/42115d37d6a46856e3dc914b54a1ce6d33b9872a/src/main/java/net/minecraftforge/network/ServerStatusPing.java
+"""
+
+from __future__ import annotations
+
+import io
+from dataclasses import dataclass
+from typing import Final, TYPE_CHECKING
+
+from mcstatus.protocol.connection import Connection
+
+VERSION_FLAG_IGNORE_SERVER_ONLY: Final = 0b1
+IGNORE_SERVER_ONLY: Final = "<not required for client>"
+
+
+if TYPE_CHECKING:
+    from typing_extensions import NotRequired, Self, TypedDict
+
+    class ForgeDataChannel(TypedDict):
+        res: str
+        """Channel name and ID (for example ``fml:handshake``)."""
+        version: str
+        """Channel version (for example ``1.2.3.4``)."""
+        required: bool
+        """Is this channel required for client to join?"""
+
+    class ForgeDataMod(TypedDict):
+        modid: str
+        modmarker: str
+        """Mod version"""
+
+    class RawForgeData(TypedDict):
+        fmlNetworkVersion: int  # noqa: N815
+        channels: list[ForgeDataChannel]
+        mods: list[ForgeDataMod]
+        d: NotRequired[str]
+        truncated: NotRequired[bool]
+
+else:
+    ForgeDataChannel = dict
+    ForgeDataMod = dict
+    RawForgeData = dict
+
+
+@dataclass
+class ForgeData:
+    fml_network_version: int
+    """Forge Mod Loader network version."""
+    channels: list[ForgeDataChannel]
+    """List of channels, both for mods and non-mods."""
+    mods: list[ForgeDataMod]
+    """List of mods"""
+    truncated: bool
+    """Is the mods list and or channel list incomplete?"""
+
+    @classmethod
+    def build(cls, raw: RawForgeData) -> Self:
+        """Build :class:`ForgeData` from raw response :class:`dict`.
+
+        :param raw: Raw forge data response :class:`dict`.
+        :return: :class:`ForgeData` object.
+        """
+        raw.setdefault("fmlNetworkVersion", 0)
+        raw.setdefault("channels", [])
+        raw.setdefault("mods", [])
+        return decode_forge_data(raw)
+
+
+def decode_optimized(string: str) -> Connection:
+    """Decode buffer UTF-16 optimized binary data from `string`."""
+    text = io.StringIO(string)
+
+    def read() -> int:
+        result = text.read(1)
+        if not result:
+            return 0
+        return ord(result)
+
+    size = read() | (read() << 15)
+
+    buffer = Connection()
+    value, bits = 0, 0
+    for _ in range(len(string) - 2):
+        while bits >= 8:
+            buffer.receive((value & 0xFF).to_bytes(length=1, byteorder="big", signed=False))
+            value >>= 8
+            bits -= 8
+        value |= (read() & 0x7FFF) << bits
+        bits += 15
+
+    while buffer.remaining() < size:
+        buffer.receive((value & 0xFF).to_bytes(length=1, byteorder="big", signed=False))
+        value >>= 8
+        bits -= 8
+    return buffer
+
+
+def decode_forge_data(response: RawForgeData) -> ForgeData:
```

Comment:
This too should be moved, but to `ForgeData.build`. It's an ideal place for parsing logic.

Reply:
Moved in 0ea7af0